[BUG] instance-manager panics (index out of range [3]) reconciling any InstanceGroup on an IPv6 EKS cluster

## Bug Description

AwsWorker.GetDNSClusterIP (controllers/providers/aws/eks.go) assumes the cluster has an IPv4 service CIDR. On an EKS cluster created with ipFamily: ipv6, EKS populates KubernetesNetworkConfig.ServiceIpv6Cidr and leaves ServiceIpv4Cidr empty. The function reads only ServiceIpv4Cidr, so on an IPv6 cluster it operates on an empty string and panics:

serviceCidr := aws.StringValue(cluster.KubernetesNetworkConfig.ServiceIpv4Cidr) // "" on IPv6
ip := strings.Split(serviceCidr, "/")[0]  // ""  -> [""]  -> ""
blocks := strings.Split(ip, ".")          // ""  -> [""]  (length 1)
blocks[3] = "10"                          // panic: index out of range [3] with length 1

GetDNSClusterIP is called unconditionally at the top of EksInstanceGroupContext.GetBootstrapArgs (controllers/provisioners/eks/helpers.go:615), which runs during Create/reconcile for both amazonlinux2 and amazonlinux2023 OS families. The panic therefore fires on every reconcile of any InstanceGroup on an IPv6 cluster, before any status is written.

## Steps To Reproduce

1. Create an EKS cluster with ipFamily: ipv6 (so ServiceIpv6Cidr is set and ServiceIpv4Cidr is empty).
2. Deploy instance-manager 0.20.3 and create any InstanceGroup (provisioner eks).
3. Observe the controller panic on reconcile; the InstanceGroup never reaches a ready state.

## Screenshots/Logs

Observed stack trace

panic: runtime error: index out of range [3] with length 1
  aws.(*AwsWorker).GetDNSClusterIP        controllers/providers/aws/eks.go:228
  eks.(*EksInstanceGroupContext).GetBootstrapArgs  controllers/provisioners/eks/helpers.go:615
  eks.(*EksInstanceGroupContext).Create   controllers/provisioners/eks/create.go:38
  controllers.HandleReconcileRequest      controllers/interface.go:42
  controllers.(*InstanceGroupReconciler).Reconcile  controllers/instancegroup_controller.go:207

## Environment

<!-- Please complete the following information -->

- Version: 0.20.3
- Kubernetes version: 1.34
- Cloud Provider: AWS

## Impact

<!-- How is this bug affecting your use of the project? -->

- [X] Blocking (cannot proceed with work)
- [ ] High (significant workaround needed)
- [ ] Medium (minor workaround needed)
- [ ] Low (inconvenience)

## Possible Solution

Make GetDNSClusterIP IPv6-aware and panic-safe:

- When ServiceIpv6Cidr is set, compute the DNS service IP as the service CIDR's network address with its last byte set to 0x0a (the ::a convention), mirroring the existing IPv4 .10 convention. Derive it via net.ParseIP on the prefix rather than string manipulation.
- In the IPv4 path, guard against a non–dotted-quad value (len(blocks) != 4) and return an empty string instead of indexing blindly, so a malformed or empty CIDR can never panic.
- Guard KubernetesNetworkConfig == nil.

The DNS IP convention is verified against a live IPv6 cluster: service CIDR fd31:8cc0:942b::/108 → kube-dns ClusterIP fd31:8cc0:942b::a, which the fix reproduces exactly.

GetBootstrapArgs already omits the --dns-cluster-ip flag when the returned value is empty (if !common.StringEmpty(clusterIP)), so returning "" for unsupported/malformed inputs is safe and requires no caller change.

--- a/controllers/providers/aws/eks.go
+++ b/controllers/providers/aws/eks.go
@@ -16,6 +16,7 @@ limitations under the License.
 package aws
 
 import (
+	"net"
 	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -217,14 +218,41 @@ func (w *AwsWorker) DescribeFargateProfile() (*eks.FargateProfile, error) {
 }
 
 func (w *AwsWorker) GetDNSClusterIP(cluster *eks.Cluster) string {
-	if cluster == nil {
+	if cluster == nil || cluster.KubernetesNetworkConfig == nil {
 		return ""
 	}
-	serviceCidr := aws.StringValue(cluster.KubernetesNetworkConfig.ServiceIpv4Cidr)
+	netCfg := cluster.KubernetesNetworkConfig
+
+	// IPv6 clusters populate ServiceIpv6Cidr and leave ServiceIpv4Cidr empty.
+	// The DNS service IP is the service CIDR's network address with its last
+	// byte set to 0x0a ("::a"), e.g. fd31:8cc0:942b::/108 -> fd31:8cc0:942b::a,
+	// mirroring the IPv4 ".10" convention below.
+	if serviceCidrV6 := aws.StringValue(netCfg.ServiceIpv6Cidr); serviceCidrV6 != "" {
+		prefix := strings.Split(serviceCidrV6, "/")[0] // remove block size
+		ip := net.ParseIP(prefix)
+		if ip == nil {
+			return ""
+		}
+		ip = ip.To16()
+		if ip == nil {
+			return ""
+		}
+		dnsIP := make(net.IP, len(ip))
+		copy(dnsIP, ip)
+		dnsIP[len(dnsIP)-1] = 0x0a // last byte "10", as is the convention
+		return dnsIP.String()
+	}
+
+	serviceCidr := aws.StringValue(netCfg.ServiceIpv4Cidr)
 	// addresses are by default assigned from either the 10.100.0.0/16 or 172.20.0.0/16 CIDR blocks.
 	// custom ranges could be blocks that are not /16 size.
 	ip := strings.Split(serviceCidr, "/")[0] // remove block size
 	blocks := strings.Split(ip, ".")
+	if len(blocks) != 4 {
+		// not an IPv4 dotted-quad (e.g. empty or malformed); avoid an
+		// index-out-of-range panic and let the caller omit --dns-cluster-ip.
+		return ""
+	}
 	blocks[3] = "10" // replace last byte in IP address with "10" as is the convention
 	return strings.Join(blocks, ".")
 }

--- a/controllers/providers/aws/eks_test.go
+++ b/controllers/providers/aws/eks_test.go
@@ -16,4 +16,17 @@ func TestClusterDns(t *testing.T) {
 	ip := awsWorker.GetDNSClusterIP(&eks.Cluster{KubernetesNetworkConfig: &eks.KubernetesNetworkConfigResponse{ServiceIpv4Cidr: &cidr}})
 	g.Expect(ip).To(gomega.Equal("172.16.0.10"))
 
+	// IPv6 cluster: ServiceIpv4Cidr is empty, ServiceIpv6Cidr is set.
+	// DNS IP is the service CIDR network address with the last byte as 0x0a.
+	cidrV6 := "fd31:8cc0:942b::/108"
+	ipV6 := awsWorker.GetDNSClusterIP(&eks.Cluster{KubernetesNetworkConfig: &eks.KubernetesNetworkConfigResponse{ServiceIpv6Cidr: &cidrV6}})
+	g.Expect(ipV6).To(gomega.Equal("fd31:8cc0:942b::a"))
+
+	// Malformed/empty service CIDR must not panic; caller omits --dns-cluster-ip.
+	empty := ""
+	ipEmpty := awsWorker.GetDNSClusterIP(&eks.Cluster{KubernetesNetworkConfig: &eks.KubernetesNetworkConfigResponse{ServiceIpv4Cidr: &empty}})
+	g.Expect(ipEmpty).To(gomega.Equal(""))
+
+	// nil network config must not panic.
+	g.Expect(awsWorker.GetDNSClusterIP(&eks.Cluster{})).To(gomega.Equal(""))
 }

